### PR TITLE
Optimize validators storage

### LINKED ISSUE #42: `valset`: Optimize VALIDATORS storage
They store 

```rust
pub struct ValidatorInfo {
    /// TODO: better name to specify this is the Tendermint pubkey for consensus?
    pub validator_pubkey: Pubkey,
    pub operator: Addr,
    pub metadata: ValidatorMetadata,
    /// The voting power in Tendermint sdk
    pub power: u64,
}
```

for each one in an `Item<Vec<ValidatorInfo>>`. I think we need to keep an eye on the footprint (max 128kb stored under one key enforced by cosmwasm-vm). Let's look at better ways of storing this, to efficiently read all (for a diff) as well as not blowing up disk space too much.

We may go to a Map based approach, but let's do research

### PR DESCRIPTION
**Pull request by maurolacy**
Closes #42.

After a lot of wrestling, added a couple integration tests to see `Item` fail.

I was determined to do this as part of `tgrade-valset` integration tests, as I think this might be useful in the future. That required adding the "tgrade" feature to cosmwasm-vm. Will publish that in a cosmwasm PR (https://github.com/CosmWasm/cosmwasm/pull/1208) and we decide what to do with it there.

~~Creating as Draft, as this is just the failure case at the moment. Will work in the proposed fix next.~~

~~Of course, tests will not pass until a new cosmwasm-vm supporting the "tgrade" feature is released.~~

### CODE REVIEW

File: contracts/tgrade-valset/src/contract.rs [hunk 1]
```diff
@@ -367,6 +370,28 @@ fn execute_slash(
...
+    // Assert admin is not found (contract not instantiated!) for this to be possible
+    let admin_res = ADMIN.get(deps.as_ref());
+    if let Err(err) = admin_res {
+        let not_found = matches!(err, StdError::NotFound { .. });
+        if !not_found {
+            return Err(ContractError::AdminError(AdminError::NotAdmin {}));
+        }
+    } else {
+        return Err(ContractError::AdminError(AdminError::NotAdmin {}));
+    }
```
Comment (maurolacy) [inline contracts/tgrade-valset/src/contract.rs:387]: There's probably a simpler / nicer way to do this.

File: .circleci/config.yml [hunk 2]
```diff
@@ -275,7 +275,7 @@ jobs:
           command: rustc --version; cargo --version; rustup --version; rustup target list --installed
       - restore_cache:
           keys:
-            - cargocache-v2-bindings:1.54.0-{{ checksum "~/project/Cargo.lock" }}
```
Comment (ethanfrey) [inline .circleci/config.yml:278]: this could be another pr, right?
or bad git diff?

File: contracts/tgrade-valset/src/contract.rs (hunk 1 above)
Comment (ethanfrey) [reply contracts/tgrade-valset/src/contract.rs:387]: we should add such method to admin.

File: contracts/tgrade-valset/Cargo.toml [hunk 3]
```diff
@@ -48,3 +48,8 @@ tg-bindings-test = { path = "../../packages/bindings-test", version = "0.6.0-alp
 derivative = "2"
 anyhow = "1"
 assert_matches = "1.5"
+
+# For integration tests
+cosmwasm-std = { version = "1.0.0-beta4", default_features = false }
```
Comment (ethanfrey) [inline contracts/tgrade-valset/Cargo.toml:53]: Can we make these all optional?
Only included with "integration" feature.

(I have a Mac m1 and won't be able to build/test this contract if these are present)

File: contracts/tgrade-valset/src/contract.rs (hunk 1 above)
Comment (ethanfrey) [reply contracts/tgrade-valset/src/contract.rs:387]: wait... if this is only for tests, we do not want this in production build at all.
how about feature-gating it with `#[cfg(features = "integration")]` and you make such a build for the wasm used in integration test? Then no worries about it as this code path isn't in production at all

File: contracts/tgrade-valset/tests/integration.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    let msg = ExecuteMsg::SimulateValidatorSet { validators };
+
+    let res: Response = execute(&mut deps, mock_env(), info, msg).unwrap();
+
+    assert_eq!(res.messages.len(), 0);
+}
+
+#[test]
+#[should_panic]
+fn check_validators_storage_breaks() {
```
Comment (ethanfrey) [inline contracts/tgrade-valset/tests/integration.rs:104]: nice to ensure we are at the limit.
1128 is big

File: contracts/tgrade-valset/tests/integration.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+#[test]
+fn test_validators_storage() {
+    let mut deps = mock_instance_on_tgrade(WASM);
+    assert_eq!(deps.required_features().len(), 2);
+
+    let info = mock_info("creator", &[]);
+
+    let validators = addrs(NUM_VALIDATORS)
+        .iter()
+        .map(|s| valid_validator(s, VALIDATOR_POWER))
```
Comment (ethanfrey) [inline contracts/tgrade-valset/tests/integration.rs:81]: one thought, what size of addresses does this use? probably not realistic ones?

I would  try bech32 encoded. This is an example on 32 bytes bech32 encoded: `wasm1zvec5uk3r3drsy2yyef3vkdrzqmt9uftyrtkexmc7fyp0r5laf7qatrjmv` (contract addresses as of wasmd 0.22)

Here is 20 bytes encoded as bech32: `wasm1r3drgfjnzev6xypkktcjkgxhdj0jfq077qyq89` ("normal" user addresses)

**Review by ethanfrey — COMMENTED** [review]
Looks good.

Please make all this integration test code feature gated.
Also ensure we have longer addresses, maybe 1 with 20 byte as bech32, other as 32 byte

File: contracts/tgrade-valset/src/contract.rs (hunk 1 above)
Comment (maurolacy) [reply contracts/tgrade-valset/src/contract.rs:387]: Good idea. Will do this tomorrow.

File: .circleci/config.yml (hunk 2 above)
Comment (maurolacy) [reply .circleci/config.yml:278]: This is lack of rebase from main I think. Will fix.

File: contracts/tgrade-valset/tests/integration.rs (hunk 5 above)
Comment (maurolacy) [reply contracts/tgrade-valset/tests/integration.rs:81]: Yeah, I was thinking about the same, as I copied these without modifying them. Let's use 32 as it is the worst case.

File: .circleci/config.yml (hunk 2 above)
Comment (maurolacy) [reply .circleci/config.yml:278]: No, this was a change I did in passing, after changing this file because of some CI error.

File: contracts/tgrade-valset/tests/integration.rs (hunk 5 above)
Comment (maurolacy) [reply contracts/tgrade-valset/tests/integration.rs:85]: Last version is using 33-byte addresses (because of the "tgrade" prefix).

File: contracts/tgrade-valset/tests/integration.rs (hunk 4 above)
Comment (maurolacy) [reply contracts/tgrade-valset/tests/integration.rs:108]: It's now 956, with 33-byte addresses. Big enough, and close to initial estimates.

File: contracts/tgrade-valset/tests/integration.rs [hunk 6]
```diff
@@ -17,21 +17,25 @@
...
-use cosmwasm_vm::{features_from_csv, Instance};
 use tgrade_valset::msg::ExecuteMsg;
 use tgrade_valset::state::ValidatorInfo;
 
 // Copied from test_helpers
 // returns a list of addresses that are set in the tg4-stake contract
 fn addrs(count: u32) -> Vec<String> {
-    (1..=count).map(|x| format!("operator-{:03}", x)).collect()
+    (1..=count)
+        .map(|x| bech32::encode("tgrade", format!("operator-{:03}", x).to_base32(), Variant::Bech32).unwrap())
```
Comment (ethanfrey) [inline contracts/tgrade-valset/tests/integration.rs]: This is getting closer, but not realistic.

The two cases I refer to are 
(1) 20 unencoded  bytes before the bech32
(2) 32 unencoded bytes before the bech32

I would replace `format!("operator-{:03}", x).to_base32()` with
`format!("20-byte-operator-{:03}", x)` which is 20 bytes

I guess we can skip the 32 byte variant for now as it is unlikely anyone uses contracts, but it would be interesting to test as a true upper limit

**Review by ethanfrey — APPROVED** [review]
One last change requested, so we get proper test numbers

File: contracts/tgrade-valset/src/contract.rs (hunk 1 above)
Comment (maurolacy) [reply contracts/tgrade-valset/src/contract.rs:387]: Forgot to remove this in the first iteration.

Then, tried to do it in a follow-up, but hit dependency / feature problems with compiling the wasm using the integration feature.

I'll post my working copy and we can see if there's a way around those problems (other than keeping / removing the entry point entirely).
